Add API for measuring event loop idle time

trevnorris opened the pull request:
This PR contains code that I've been working on for a while as part of some research I've been doing at work, and decided it would be good to submit this for review after seeing @indutny's Node.js PR https://github.com/nodejs/node/pull/32018.

The main reasons for writing this patch are:
- Have very accurate measurements of idle time spent in the kernel's event provider.
- Allow retrieval of the idle time from another thread to allow more accurate time-series data collection (instead of only collecting the data when the event loop can process a callback).

A lot is explain in the commit message below, and I'll be happy to explain in detail why the implementation was done this way. For example, this approach was a conclusion I came to last year after researching whether it was possible to infer the state of the application based solely on event loop metrics. In order to achieve that goal it's necessary to have the level of accuracy this patch provides.

There are a few things I'm unsure about and would like feedback on, like how I've implemented atomics, and I'll need some assistance with the Windows implementation.

/cc @bnoordhuis @indutny @saghul 

Commit message:

Add API for measuring event loop idle time

The API addition `uv_metrics_idle_time()` is a thread safe call that
allows the user to retrieve the amount of time the event loop has spent
in the kernel's event provider (i.e. poll). It was done this way to
allow retrieving this value without needing to interrupt the execution
of the event loop. This option can be enabled by passing
`UV_METRICS_IDLE_TIME` to `uv_loop_configure()`.

One important aspect of this change is, when enabled, to always first
call the event provider with a `timeout == 0`. This allows libuv to know
whether any events were waiting in the event queue when the event
provider was called. The importance of this is because libuv is tracking
the amount of "idle time", not "poll time". Thus the provider entry time
is not recorded when `timeout == 0` (the event provider never idles in
this case).

While this does add a small amount of overhead, when enabled, but the
overhead decreases when the event loop has a heavier load. This is
because poll events will be waiting when the event provider is called.
Thus never actually recording the provider entry time.

Checking if `uv_loop_t` is configured with `UV_METRICS_IDLE_TIME` always
happens in `uv__metrics_set_provider_entry_time()` and
`uv__metrics_update_idle_time()`. Making the conditional logic wrapping
each call simpler and allows for instrumentation to always hook into
those two function calls.

Add struct `uv__metrics_loop_t` as a location for future metrics API
additions.

Add struct `uv__loop_internal_fields_t` as a location for future
additions to `uv_loop_t` while also maintaining v1.x compatibility. A
pointer to this struct has been placed in the `uv_loop_t::unused[1]`
space to allow for fast retrieval and maintaining ABI compatibility with
v1.x.

Currently `uv__loop_internal_fields_t` only contains both the
`uv__metrics_loop_t` and `flags` field. The reason for adding the
`flags` field is because the same field was never added to
`UV_LOOP_PRIVATE_FIELDS` in Windows as it was in Unix, and adding that
field would break ABI.

The idea for creating a struct and attaching it to `uv_loop_t` for
future API enhancements was taken from a comment by bnoordhuis in:
https://github.com/libuv/libuv/issues/2506#issuecomment-540050665

Tests and additional documentation has been included.

santigimeno commented on src/unix/aix.c:
```diff
@@ -214,6 +215,16 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
+  if (user_timeout != 0) {
+    uv__metrics_set_provider_entry_time(loop, uv_hrtime());
+  }
+
```
Style nit: remove braces in 1-line conditionals

santigimeno commented on src/unix/kqueue.c:
```diff
@@ -202,6 +203,16 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
+   */
+  if (user_timeout != 0) {
+    uv__metrics_set_provider_entry_time(loop, uv_hrtime());
+  }
```
Style nit: remove braces

santigimeno commented on src/unix/linux-core.c:
```diff
@@ -281,6 +282,16 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
+   */
+  if (user_timeout != 0) {
+    uv__metrics_set_provider_entry_time(loop, uv_hrtime());
+  }
```
Also here

santigimeno commented on src/unix/loop.c:
```diff
@@ -68,6 +69,13 @@ int uv_loop_init(uv_loop_t* loop) {
   if (err)
     return err;
 
+  mloop = (uv__metrics_loop_t*) uv__calloc(1, sizeof(uv__metrics_loop_t));
```
This call might fail. Take it into account.

trevnorris replied:
You are correct. I've moved this to near the top of the function and have it return `UV_ENOMEM` if it fails.

santigimeno commented on src/unix/core.c:
```diff
@@ -1517,3 +1526,55 @@ void uv_sleep(unsigned int msec) {
+
+
+void uv__metrics_update_idle_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop = uv__metrics_get_mloop(loop);
```
Style nit: we usually separate declaration from assignment.

saghul commented:
The performance impact can probably be mitigated if the machinery is disabled by default and it's only initialized with a call to `uv_loop_configure`, WDYT?

As for the general approach, I'm not an expert, but arent tools like perf and eBPF capable of gathering this insight, without instrumenting the application?

Regarding the specific metric introduced: in what context is it useful? The way libuv operates it will have 3 types of values:

* 0: can't block, there are pending callbacks to be called.
* -1: block indefinitely
* N: closest timer

What kind of application would be interested in these?

trevnorris commented:
@saghul 
> The performance impact can probably be mitigated if the machinery is disabled by default and it's only initialized with a call to `uv_loop_configure`, WDYT?

That sounds like a reasonable solution.

> As for the general approach, I'm not an expert, but arent tools like perf and eBPF capable of gathering this insight, without instrumenting the application?

This is meant to be used in a production environment where the data can be collected at mass scale. The overhead is low enough that it has no noticeable impact on a (Node.js) process. It also serves as a single interface that is platform agnostic.

> Regarding the specific metric introduced: in what context is it useful?

Allow me to dive into some of my research. Last year I began to investigate:
1. How to scale Node.js applications, now that the classical scaling rules (largely based on a process's CPU usage) no longer apply in a world of Worker threads.
2. How to understand the overall state of a process based on metrics only collected from the event loop.

Note: From here on the percentage of idle time spent in the event loop's event provider will simply be referred to as the "event loop utilization" or ELU.

To better understand the problem scope I began running simulations of (Node.js) HTTP server request/response characteristics based on different workloads, and did comparisons of the number of requests each workload could support. Then I did a comparison of the request data to the process's CPU usage and the ELU. After doing a lot of analysis on several dozen types of workloads I found a correlation between these three data points. This research, and the analysis presented below on the ELU, come from the patch in this PR.

The following two graphs represent two different server workloads. The X-axis is the number of connections to the server. The right Y-axis is the number of responses (blue line). The left Y-axis is the CPU% (yellow line) and the ELU (red line).

Notice that the graph shows "req/period" and not "req/second". This is because for a complex applications the number of requests per second isn't as reliable a metric. So some measurements are the number of requests per 5 or 10 seconds. Simply using "req/period" is a generic way of labeling this.

<p float="left">
<img src="https://user-images.githubusercontent.com/289174/75906248-a2f37b00-5e03-11ea-9b50-ad12ba09ab56.png" alt="req_per_period_chart1" width="300" />
<img src="https://user-images.githubusercontent.com/289174/75906257-a555d500-5e03-11ea-948d-fd03dc02946f.png" alt="req_per_period_chart2" width="300" />
</p>

The differences in the response curves based on the workload type is a discussion on its own, but that's for another time.

Instead we'll focus on the fact that in these two cases the ELU almost perfectly matches the CPU. This is because in both of these workloads V8 is able to GC immediately, instead of needing to do any off-thread or idle-time cleanup. Also notice that both CPU and ELU closely match the response curve. And both curves closely match the req/period curve. I found this to generally be true, except when doing things like HTTP request pipelining.

Now for the next example:

<img src="https://user-images.githubusercontent.com/289174/75909611-a68a0080-5e09-11ea-8fd9-7dd1b805a1fa.png" alt="req_per_period_chart3" width=300 />

This use case causes the GC to be constantly doing cleanup off-thread and during idle-time. The CPU usage maxes out at around 125%. If we were to scale the application soley based on CPU usage (which is how most Node.js applications are scaled), let's say at 80%, then we could be scaling to early. At 80% CPU usage ELU% is only at 61% and the server is serving 369 req/period, of the maximum 710 req/period that it's capable of. This translates to scaling the application when it's only at ~51% of its actual maximum capacity.

Note that in this graph it shows the total req/period increasing slightly after the ELU has reached 100%. This is because of V8 GC and code optimization specifics.

To recap the above, the ELU has the following advantages over the usual CPU usage:

1. Its range is only between [0, 1]. Making it simpler to set application scaling rules.
2. It only reflects the amount of time the event loop is busy. Thus giving a more accurate indicator of how much additional work the process is capable of handling.
3. It is a cross-platform way to monitor every thread's event loop. This is especially relevent to Node's Worker threads. Node processes in the future could easily be running at 300% CPU usage and still be completely healthy.

> The way libuv operates it will have 3 types of values:

Sorry, I'm not sure how the three timeout values are applicable.

trevnorris commented:
@saghul I've added the loop configuration option `UV_LOOP_IDLE_TIME`. Also added it to the Windows side as well since support is planned before this PR lands.

santigimeno commented on src/unix/loop.c:
```diff
@@ -36,6 +37,15 @@ int uv_loop_init(uv_loop_t* loop) {
+  mloop = (uv__metrics_loop_t*) uv__calloc(1, sizeof(uv__metrics_loop_t));
+  if (mloop == NULL)
+    return UV_ENOMEM;
+
```
Maybe allocating `uv__metrics_loop_t`  only if/when `UV_LOOP_IDLE_TIME` is set?

trevnorris replied:
From @bnoordhuis comment in https://github.com/libuv/libuv/issues/2506#issuecomment-540050665:

> I think that can be addressed by malloc-ing memory.

Since that's basically what I'm doing with `uv__metrics_loop_t`, maybe it's not a bad time to create a `struct` that will hold all future would-be ABI changes. Something like `uv__loop_internal_t`?

@bnoordhuis @saghul Thoughts?

santigimeno commented on test/test-metrics.c:
```diff
@@ -0,0 +1,112 @@
+static void metrics_thread_timer_cb(uv_timer_t* handle) {
+  mthread_t* mthread = (mthread_t*) handle->data;
+  uint64_t pt = uv_metrics_idle_time(&mthread->mloop);
+  uint64_t ptt = uv_metrics_idle_time(handle->loop);
```
Style nit: separate declaration from assignment

santigimeno commented on test/test-metrics.c:
```diff
@@ -0,0 +1,112 @@
+}
+
+static void metrics_thread_routine_cb(void* arg) {
+  mthread_t* mthread = (mthread_t*) arg;
```
Also here

santigimeno commented on test/test-metrics.c:
```diff
@@ -0,0 +1,112 @@
+}
+
+static void metrics_thread_async_cb(uv_async_t* handle) {
+  mthread_t* mthread = (mthread_t*) handle->data;
```
And here

santigimeno commented on test/test-metrics.c:
```diff
@@ -0,0 +1,112 @@
+
+  /* Technically should be able to check if >= 100 ms, but the timer might run
+   * slightly early. So check if >= expected time - 1 ms.
+   */
```
Nit: may check that the timer cb has been called.

trevnorris replied:
Good call.

santigimeno commented on src/unix/core.c:
```diff
@@ -1517,3 +1526,68 @@ void uv_sleep(unsigned int msec) {
+  if (entry_time > 0)
+    idle_time += uv_hrtime() - entry_time;
+  return idle_time;
+}
```
This should handle the case where `UV_LOOP_IDLE_TIME` is not set by returning `UV_EINVAL`. Maybe changing the signature to something like: `int uv_metrics_idle_time(uv_loop_t* loop, uint64_t* idle_time)` might work.

trevnorris replied:
I think simply returning `0` is enough. The API documentation says it only starts counting once the flag has been set, so IMO it's within reason that no time has been collected before being set.

santigimeno commented on src/unix/core.c:
```diff
@@ -1517,3 +1526,68 @@ void uv_sleep(unsigned int msec) {
+  uint64_t now;
+
+  if (!(loop->flags & UV_LOOP_IDLE_TIME))
+    return;
```
Maybe asserting here as it should not be called if `UV_LOOP_IDLE_TIME` is not set.

trevnorris replied:
But it is always called. To simplify the logic at each call site I placed the check in the function. Otherwise all 14 call locations would need to be wrapped in this same conditional.

saghul commented:
@trevnorris Thanks for the thorough explanation, I think I get what you're after here. 👍 

> > The way libuv operates it will have 3 types of values:
> 
> Sorry, I'm not sure how the three timeout values are applicable.

Disregard that, I'm nott sure where I was going there... 🙈

mmarchini commented:
> As for the general approach, I'm not an expert, but arent tools like perf and eBPF capable of gathering this insight, without instrumenting the application?

While this is true, the application/libraries still need some level of instrumentation for eBPF to be used properly. Inlined functions, lack of debug symbols and frame pointers omitted usually hurt eBPF observability. Also, BPF is only available on Linux, and is complicated to use on FaaS or containerized platforms.

A while ago I suggested re-adding DTrace-style tracepoints to libuv (#2209) to make the event loop more observable. But I like this approach better, as it doesn't require users to learn and setup a system-level observability tool and can be used regardless of underlying stack or operating system. And for users who need those system level tools, this approach improves the information they can get, becaue now it's possible to hook those tools to uv__metrics_update_idle_time.

indutny commented on src/uv-common.h:
```diff
@@ -323,4 +323,13 @@ void* uv__malloc(size_t size);
 void uv__free(void* ptr);
 void* uv__realloc(void* ptr, size_t size);
 
+typedef struct {
```
Nitpick, but I think we declare similar structures using `struct uv_metrics_s {...}` and `typedef struct uv_metrics_s uv_metrics_t`

trevnorris replied:
I knew this was the case in `include/uv.h` but wasn't sure about the code in `src/`. So I took my example from `uv__allocator_t` in `src/uv-common.cc`. But after doing a quick `grep` I see that pattern is still usually the case here. I'll change it to follow that pattern.

trevnorris commented on src/unix/aix.c:
```diff
@@ -214,6 +215,20 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
+  if (loop->flags & UV_LOOP_IDLE_TIME) {
+    user_timeout = timeout;
+    timeout = 0;
```
Agreed this code is confusing, but after many attempts also turned out to be the implementation that was the least invasive. I'll try to explain:

1 - Backup the user's timeout value into `user_timeout`.
2 - Set `timeout = 0`.
3 - Run `epoll_wait` with `timeout == 0`.
4 - When it returns, set `timeout = user_timeout` and set `user_timeout = -2` to indicate that the first pass has already run.
5 - if `nfds > 0` then:
    5.1 - If an event is processed (meaning `w->cb()` is called) then end accumulating idle time
    5.2 - If a signal is received, and the user has a signal watcher set, then end accumulating idle time
6 - If after processing all `fd` and `nevents == 0` and the original `timeout == -1 || timeout > 0` then reenter `epoll_wait`
7 - If after processing all `fd` and `nevents > 0` then return immediately

Everything after step 4 happens anyway. So it just forces the first call to poll to return immediately to see if any events were waiting when poll was called (since there's no other way to determine if there were).

trevnorris replied:
@indutny Is this explanation clear? Since it's duplicated in all poll implementations I didn't add many inline comments. If you think that's necessary then I'll be happy to.

indutny replied:
It is almost clear. Could you add the `uv__metrics_update_idle_time` and `uv__metrics_set_provider_entry_time` to the list of your items?

What I'm worried about is that `uv__metrics_set_provider_entry_time` is called before `3`, and `uv__metrics_update_idle_time` is called before `6`. So when we get into `7` and eventually back to `3` the `entry` is zero. Does it sound right?

trevnorris replied:
@indutny I've redone the logic here. Using `user_timeout == -2` to control the loop was a stupid way of preventing the need to use one additional variable. I've changed the implementation and  hopefully made it more apparent. PTAL.

trevnorris replied:
Resolving this since the logic has since changed.

trevnorris commented:
Review suggestions have been addressed.

Is there anyone willing to give tips on the Windows implementation? The first question I need to answer is if `GetQueuedCompletionStatusEx` (used in `uv_poll`) is similar to all unix implementations. In that `GetQueuedCompletionStatusEx` returns when there are events to be processed, but it is still libuv's job to process those events. Or does `GetQueuedCompletionStatusEx` process the events for the user then return? I can't tell when the user's callback is supposed to be called.

The other main thing is to see if there are platform-agnostic atomics. I can switch to using a mutex lock if necessary, but I'd like to prevent that overhead if possible. If anything guess I could `#ifdef` and only use atomics on supported platforms and use a mutex on the other. @bnoordhuis have any suggestions here?

trevnorris commented:
My worry about `GetQueuedCompletionStatusEx` is this from the documentation:

> Retrieves multiple completion port entries simultaneously. **It waits for pending I/O operations that are associated with the specified completion port to complete.**

Which leads me to believe the event's callback is called before returning. This is a serious hindrance because it's then impossible to tell exactly when the event was received. I'll investigate the I/O Completion Port API further, but a pointer in the right direction would save me a lot of time.

santigimeno commented:
> Is there anyone willing to give tips on the Windows implementation? The first question I need to answer is if GetQueuedCompletionStatusEx (used in uv_poll) is similar to all unix implementations. In that GetQueuedCompletionStatusEx returns when there are events to be processed, but it is still libuv's job to process those events. Or does GetQueuedCompletionStatusEx process the events for the user then return? I can't tell when the user's callback is supposed to be called.

/cc @libuv/windows

bzoz commented:
The user callbacks are called by [`ran_pending = uv_process_reqs(loop);`](https://github.com/libuv/libuv/blob/e8644693ea2449360e2f06c8b5792e923ab4aeef/src/win/core.c#L524).

trevnorris commented:
Added a commit that will switch to using a `uv_mutex_t` if the GNU C atomic extension isn't available. Don't love it, but there's a better solution then I'll stick with this.

trevnorris commented:
Did initial Windows support (much thanks @bzoz for pointing that out). Still need to do some testing.

Also, the implementation of the internal metrics functions is almost identical. Except for the atomic operations. I'm going to benchmark the performance difference between using the atomics and the mutex. Depending on the difference, I might just drop the atomics part completely and unify the implementations.

trevnorris commented:
Here are the changes I've made:

* Squashed all commits and rebased on latest v1.x
* Removed using atomics and unified implementation using `uv_mutex_t`.
* Added Windows support (thanks @bzoz for the tip)
* Made it so the idle time APIs are always called. They'll return early if not configured to collect idle time metrics, but did this so the functions could be instrumented regardless (@mmarchini that sound good?).

Think it's almost ready. Would appreciate another round of reviews.

trevnorris commented:
So, while I was testing on Windows I came to learn that Windows doesn't have a `flags` field on its `uv_loop_t`. I assume adding it to `UV_LOOP_PRIVATE_FIELDS` in `include/uv/win.h` will break ABI?

One possible solution is that I put whether it's enabled on the `uv__metrics_loop_t`. But it's not explicitly documented that you can't run `uv_loop_configure` on a `uv_loop_t` prior to running `uv_loop_init`. In that case I guess I could return some type of `UV_E*` error.

@saghul  Any thoughts on this?

trevnorris commented:
Recent changes:
* Added `uv__loop_internal_fields_t` to store future expansions to `uv_loop_t` without breaking ABI. This idea was presented by @bnoordhuis in https://github.com/libuv/libuv/issues/2506#issuecomment-540050665. This started from the fact that `uv_loop_t` in Windows does not have the `flags` field.
* Renamed a few variables and function names to make more sense.

Related tests pass on Windows and Linux. Last problem I'm facing is running the test with `valgrind` causes the metrics tests to fail. So far haven't been able to figure out why but will look further into it.

trevnorris commented:
Alright. Fixed the `valgrind` test failures and cleaned up a couple things. Should be ready to go.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1790/

santigimeno commented on test/test-metrics.c:
```diff
@@ -0,0 +1,159 @@
+  uv_timer_init(uv_default_loop(), &timer);
+
+  uv_timer_start(&timer, timer_noop_cb, 100, 0);
+
```
Nit: check return values of the uv_ methods here and in the other tests

trevnorris replied:
Fixed them all.

trevnorris commented on src/unix/linux-core.c:
```diff
@@ -281,7 +283,21 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
+     * will return early if the loop isn't configured with UV_METRICS_IDLE_TIME.
+     */
+    if (timeout != 0)
+      uv__metrics_set_provider_entry_time(loop);
```
That's the way it would always happen. It just wasn't as apparent before.

Reason why idle time is not recorded when `timeout == 0` is because at no time is poll actually idling. It's requesting events from the kernel's event queue then immediately returning.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1793/

trevnorris commented:
Only tests that are failing are on zos:
```
not ok 158 - metrics_idle_time
# exit code 3
# Output from process `metrics_idle_time`:
# pt: 0
# Assertion failed in ../test/test-metrics.c on line 52: pt >= 80000000
# CEE5207E The signal SIGABRT was received.
not ok 159 - metrics_idle_time_thread
# exit code 3
# Output from process `metrics_idle_time_thread`:
# pt: 0   ptt: 0
# Assertion failed in ../test/test-metrics.c on line 74: pt >= (100000000 * mthread->thread_timer_count) - 20000000
```
Conceptually this should be impossible, and not something I can debug on my own. There a zos group that could help me out with this?

santigimeno commented:
I think that might be @libuv/ibmi

kadler commented:
Sorry, but us in @libuv/ibmi only know IBM i, not z/OS. Not sure who the z/OS maintainers are. IBM i is not yet in the CI system (but we're working on it).

trevnorris commented:
Made a few changes to the PR:
* Rebased on latest `v1.x`
* Squashed all but the "Add printf()` statements [...]" commit (will remove that commit before it lands)
* Change the pointer of `uv__metrics_loop_t` in `uv__loop_internal_fields_t` to a field. No reason to allocate memory twice.

@kadler Do you know someone that could test IBM i and let me know if it works? Also @jBarz was the one that did the current z/OS implementation (in 011e02e3e5b). He still work at IBM?

At the minimum I'd like to get the approval from @bnoordhuis and/or @saghul. Also @piscisaureus or @vtjnash.

santigimeno commented:
/cc @richardlau ?

cjihrig commented:
The zos team is @libuv/zos.

kadler commented:
@dmabupt can you please test these changes?

miladfarca commented:
It seems like this PR was never ported to zos (`unix/os390.c`): https://github.com/libuv/libuv/commit/70002c80bf1447fad605b61d9d6e83e9d0c1a6bc
```c
 if (timeout == 0)
        return;
```
Above block doesn't exist in current zos implementation of `unix/os390.c` hence the while loop keeps getting taken. As `timeout = 0` , `uv__metrics_set_provider_entry_time` never gets executed and `provider_entry_time` remains as `0`. Hence final result is also a 0.

Not sure why above PR wasn't added to zos but adding the changes passes the 2 above tests.

dmabupt commented:
> @dmabupt can you please test these changes?

@kadler , Seems that the new feature work fine on IBMi PASE, (althought the output once made me thought that something wrong happened).
```
ok 160 - metrics_idle_time
# Output from process `metrics_idle_time`:
# pt: 112561479
ok 161 - metrics_idle_time_thread
# Output from process `metrics_idle_time_thread`:
# pt: 129074486   ptt: 129035437
# pt: 259156344   ptt: 258111347
# pt: 389257352   ptt: 388149161
ok 162 - metrics_idle_time_zero
# Output from process `metrics_idle_time_zero`:
# pt: 0
```

trevnorris commented:
@miladfarca Thank you much for investigating this. Definitely not something I could have found on my own.

I'll port that change here and reorder the commits to make it the first in this PR.

@dmabupt Sorry about that. I added a debug commit that always printed those values so I could get some info. Will be removing that commit before landing. Thanks for testing.

richardlau commented:
> 
> 
> It seems like this PR was never ported to zos (`unix/os390.c`): [70002c8](https://github.com/libuv/libuv/commit/70002c80bf1447fad605b61d9d6e83e9d0c1a6bc)
> 
> ```c
>  if (timeout == 0)
>         return;
> ```
> 
> Above block doesn't exist in current zos implementation of `unix/os390.c` hence the while loop keeps getting taken. As `timeout = 0` , `uv__metrics_set_provider_entry_time` never gets executed and `provider_entry_time` remains as `0`. Hence final result is also a 0.
> 
> Not sure why above PR wasn't added to zos but adding the changes passes the 2 above tests.

cc FYI @IgorTodorovskiIBM

trevnorris commented:
I have ported the commit mentioned by @richardlau and added it as the first commit in this PR. I've also rebased on latest `v1.x` and removed the debug commit.

@santigimeno Mind kicking off CI one more time?

santigimeno commented:
Sure. CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1801/

trevnorris commented:
Not sure why it's giving a 404 for `libuv-test-commit-windows` and `libuv-test-commit-windows-cmake`, but that patch does appear to have fixed the z/OS issue (thanks again @richardlau).

The last bit I'd like confirmation on is using `uv_loop_t.unused[1]`. I was thinking we could simply rename `unused` to `internals`, or something like that. I don't believe using that field will be a problem. Best I can tell `sizeof(void*) >= sizeof(unsigned int)` on all supported platforms. Any thoughts on this?

Also, is there anyone who doesn't like this change? Adding `uv__loop_internal_fields_t` helps solidify a long term method of staying on `v1.x` while being able to add new features.

richardlau commented:
(Just for the record the commit that fixed the z/OS issue was found by @miladfarca.)

trevnorris commented:
Rebased on latest `v1.x` and added a commit that changes using `active_reqs.unused[1]` to simply `internal_fields`.

indutny commented on src/unix/aix.c:
```diff
@@ -214,7 +216,21 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
+     * will return early if the loop isn't configured with UV_METRICS_IDLE_TIME.
+     */
+    if (timeout != 0)
+      uv__metrics_set_provider_entry_time(loop);
```
👍

trevnorris commented:
Ping @saghul or @bnoordhuis

trevnorris commented:
I've reordered the commits. Now the addition of `uv__loop_internal_fields_t` has been placed before the metrics patch. I believe this will make bisecting and potential reverts easier. Also renamed a few variables to make things more clear.

@santigimeno Mind kicking off another CI to make sure I didn't accidentally mess something up?

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/

santigimeno commented:
@libuv/collaborators it would be great having more eyes look into this as it's touching loop internals. Thanks.

bnoordhuis commented:
Hi Trevor, long time no see. :-)

The API looks more complex than I would expect it to be. The goal is to obtain the precise time libuv spends in epoll_pwait/kevent/etc.?

I'd say that can be simplified to two simple callbacks, either global or per-loop. Metrics collection shouldn't be libuv's job, just the means to enable it.

Strawman:
```c
// uv.h
typedef void (uv_before_io_poll_cb)(uv_loop_t*);  // exact prototype TBD
typedef void (uv_after_io_poll_cb)(uv_loop_t*);  // exact prototype TBD

int uv_set_before_io_poll_cb(uv_loop_t* loop, uv_before_io_poll_cb cb);
int uv_set_after_io_poll_cb(uv_loop_t* loop, uv_after_io_poll_cb cb);

// linux-core.c etc.
void uv__io_poll(uv_loop_t* loop, int timeout) {
  // ...

  if (loop->before_io_poll_cb != NULL)
    loop->before_io_poll_cb(loop);

  nfds = epoll_pwait(/* ... */);

  if (loop->after_io_poll_cb != NULL)
    loop->after_io_poll_cb(loop);

  // ...
}

// libuv consumer
void before_io_poll(uv_loop_t* loop) {
  uint64_t now = uv_hrtime();
  // ...
}

void after_io_poll(uv_loop_t* loop) {
  uint64_t now = uv_hrtime();
  // ...
}
```
Potential flaws:

* doesn't compose, there can be only one set of callbacks (fixable)

* with composability you'll get questions from people on how to ensure they're _really_ the last or first callback :-/

* embedders may want to influence poll behavior from the before_io_poll_cb, e.g, always poll with a zero timeout (fixable but see composability)

trevnorris commented:
@bnoordhuis Really has been a long time.  :)

>  The goal is to obtain the precise time libuv spends in epoll_pwait/kevent/etc.?

More specifically it's an attempt to measure the amount of **idle time** in the event provider (e.g. `epoll_pwait`, etc.). Below is a timeline of the event loop making event provider requests and time spent processing those requests; and when events are received by the event queue.

<pre style="line-height:1;"><code>                                ╭──────╮
                                │ idle │
      ╓───┬───────────┬─────────┼──────┼───╖
event ║   │           │         ┝━━━━━━┥   ║
loop  ║   │           │         ┝━━━━━━┥   ║
      ╟─┬─┴───┬───┬───┴─────────┴──────┼───╢
event ║ │     │   │                    │   ║
queue ║ │     │   │                    │   ║
      ╙─┴─────┴───┴────────────────────┴───╜</code></pre>

Here we can see that regardless of when events are placed in the event queue they are not received by the event loop until after it has completed processing all previous events. In these cases the event loop was never actually idle during the event provider request. Since the purposed of the metrics is to measure the idle time we need to be able to determine if the event provider request was ever actually idle during its execution.

In order to make this measurement more precise, part of this PR always makes the initial event provider request (i.e. call to `epoll_pwait`) with a `timeout == 0` (I'm doing it this way because there isn't another method to get the necessary information). Since the event provider request returns immediately we can surmise whether a subsequent request will contain some idle time. Afterwards it'll reset `timeout` and resume execution as normal. Potentially making the event provider request again if `timeout > 0` and no events were waiting in the event queue.

> I'd say that can be simplified to two simple callbacks, either global or per-loop. Metrics collection shouldn't be libuv's job, just the means to enable it.

I don't disagree with you, but I also couldn't find a simpler method to collect metrics with the precision necessary to support the research I explained in https://github.com/libuv/libuv/pull/2725#issuecomment-594752268 using a simple callback API. For example, if all time in the event provider was considered idle time then idle time is always `> 0`. Even if there are always waiting to be processed in the event queue.

trevnorris commented:
> with composability you'll get questions from people on how to ensure they're really the last or first callback :-/

@bnoordhuis I forgot to mention that another reason for my API decision is because I wanted to remove the effect of users' observation. The timing is sensitive and as soon as multiple users are allowed to run callbacks the collected metrics all interfere w/ each other. In which case the cat always ends up dead.

I'm not opposed to changing the implementation, and will be happy to do so if we can find one that 1) has the necessary precision and 2) doesn't cause multiple users to interfere with each other.

jasnell commented:
@trevnorris ... this is excellent work and I'm definitely excited to see this land. I don't want to side track any of the discussion happening here but I did want to ask if you also had plans for a corresponding plan to expose the new metric in Node.js?

trevnorris commented:
@jasnell I do believe this information should be exposed in Node. I figured `performance.idleTime()` would be an appropriate API, and the implementation is simple if Node just wants to return the same value as `uv_metrics_idle_time()`. In addition, one reason I made this API thread safe is for easier Worker thread support. If this lands, I'll go into greater detail in a Node PR.

trevnorris commented:
Ping @libuv/collaborators. Any further thoughts on this?

jasnell commented:
@trevnorris ... the one thing that I did notice on this is that using `uv_loop_configure()`, it is possible to switch on the metrics collection but it does not appear possible to turn it *off* and I wanted to confirm that it was intentional. The idea here is that the metrics collection would be enabled either at the start of the event loop operation or turned on at some point after and left on?

The decision on that would impact how we design an API for this in Node.js. That is, if it's possible to toggle it on and off over the course of a loop lifetime, then providing that ability in the Node.js API would be desirable. If it's meant to be turned on and left on, then the API becomes much easier on the Node.js side.

trevnorris commented:
@jasnell Since `uv_loop_configure()` doesn't provide a way to disable a configuration I didn't bother to handle a situation where it would be disabled after being enabled. So unless libuv adds something like `uv_loop_unconfigure()` it will probably only be allowed to stay on once enabled.

jasnell commented:
that definitely makes things easier, and it should be low enough perf impact that we could decide just to have it always on in node.js. Thanks @trevnorris :-) ... looking forward to seeing this land

trevnorris commented:
@bnoordhuis I'd like to get some more feedback from you on this PR and how it can be improved and accepted.

In https://github.com/nodejs/node/pull/33138#issuecomment-621714442 you mentioned using hooks around poll. In my PR I've intentionally had `uv__metrics_set_provider_entry_time()` run around poll, but also on the condition that `timeout != 0`. Reason for that is because whenever `timeout == 0` then poll is never idle.

Likewise it will only run just before the first event is about to be processed. This is to get around the fact that polling may be interrupted, but the file descriptor will only be disarmed and poll is re-entered. As explained the the comment:
> File descriptor that we've stopped watching, disarm it.

The placement and conditions here would be the most idle for any callback hooks you think would be worth adding.

bnoordhuis commented:
I'm afraid I'm a bit short on time right now but I'll try to come back to this early next week.

trevnorris commented:
ping @bnoordhuis

jasnell commented:
Ping @bnoordhuis @cjihrig @saghul ... it would be fantastic if we could move this forward.

jasnell commented:
@indutny @santigimeno ... anyone.... would it be possible to get an update on this?

jasnell commented:
@trevnorris... It's not ideal, but any interest in floating this patch in nodejs since it doesn't seem to be making progress here for some unknown reason?

trevnorris commented:
@jasnell I don't have any concerns of nodejs floating this patch. We use it internally at NodeSource so I have to keep it rebased on all the libuv tags that node uses. Not just the v1.x branch.

jasnell commented:
Let's do that then... I'll open an PR that (at least temporarily) floats this change in nodejs/node master, with the idea that if/once it lands in a libuv release we can drop that. The only caveat is that every time libuv is updated, we'll need to reapply the floated patch. That will allow us to move forward with the idle time API as an experimental addition to core.

richardlau commented:
@jasnell won’t floating a patch that is used by Node.js core break when using a shared libuv (like several Linux distros do)?

jasnell commented:
Yup. Which is why I'd much prefer this to land here! But it would at least allow us to make progress

jasnell commented:
Ping @cjihrig @indutny @santigimeno @saghul ... is there anything @trevnorris and I can do to get this moving forward?

vtjnash commented on include/uv.h:
```diff
@@ -247,7 +247,8 @@ typedef struct uv_utsname_s uv_utsname_t;
 
 typedef enum {
-  UV_LOOP_BLOCK_SIGNAL
+  UV_LOOP_BLOCK_SIGNAL = 1,
+  UV_METRICS_IDLE_TIME = 2
 } uv_loop_option;
```
```suggestion
  UV_LOOP_BLOCK_SIGNAL = 0,
  UV_METRICS_IDLE_TIME = 1
} uv_loop_option;
```

trevnorris commented:
@vtjnash Thanks for pointing that out. I've made the change and rebased on latest `v1.x`.

> I think this provides a reasonable framework that other times could be added later, if there was value, or that this can be changed to be more accurate, if other data becomes available.

That hits exactly why this PR was designed the way it was. Internally I've extended this to collect various other metrics, but before contributing those I wanted to make sure this approach would be accepted.

vtjnash commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1962/ (will plan to push to v1.x when green)

vtjnash commented:
Looks like Windows is not actually entirely happy with this
```
17:11:41 not ok 166 - metrics_idle_time_thread
17:11:41 # exit code 3
17:11:41 # Output from process `metrics_idle_time_thread`:
17:11:41 # Assertion failed in C:\workspace\libuv-test-commit-windows-cmake\test\test-metrics.c on line 61: `pt >= (100000000 * mthread->thread_timer_count) - 20000000` (3170200 >= 80000000)
```

vtjnash commented:
Reading closer, 100ms does seem a bit aggressive of a measurement timeout. Perhaps 1.5 second, with permitted measurements being ±1 second of that might be typically more reliable?

trevnorris commented:
@vtjnash I'm fine with that if everyone else is. Just didn't want a single test to take too long to run. I'll make the change now.

trevnorris commented:
@vtjnash I've added a commit with the changes to the test. I've done a 1 sec timer timeout and check the idle time to ±500 ms. Also in the timer's callback I spin for 600 ms to possibly force the idle time return over the ±500 ms leeway.

If the changes to the test are acceptable then I'll squash the last commit.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1966/

trevnorris commented:
Bummer. Same failure. I'm now managing to reproduce it on my Windows VM. I'm confused why this is happening on a spawned thread. Even when I only spawn a single thread, and use the noop timer callback, it still shows an extremely low idle time. Will give this another look tomorrow.

trevnorris commented:
@vtjnash @santigimeno I've updated and cleaned up the PR. Tests should be passing on Windows. The bug came from me overlooking this comment in `uv__poll()`:
```
      /* GetQueuedCompletionStatus can occasionally return a little early.
       * Make sure that the desired timeout target time is reached.
       */
```
The result was that the internal tracking mechanism for when poll was entered was being unexpectedly reset. So placement of `uv__metrics_update_idle_time()` has been changed to accommodate that (along with a code comment explaining for the reasoning behind the change).

Should be ready for CI.

vtjnash commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1969/

trevnorris commented:
I forgot to remove the `struct mthread_t` from `test/test-metrics.c` after rewriting the tests. So just removed that and rebased it on latest `v1.x`. No other changes.

vtjnash commented:
Merge commit CI: https://ci.nodejs.org/job/libuv-test-commit/1973/

trevnorris commented:
@vtjnash Thanks a lot for helping this make it through.

addaleax commented:
The documentation here should probably mention the unit of the returned value as well
